code/modules/projectiles/guns/energy/capacitor.dm

```diff
@@ -0,0 +1,163 @@
+		/decl/material/solid/gemstone/diamond = MATTER_AMOUNT_TRACE
+	)
+
+	var/capacitor_type = /obj/item/stock_parts/capacitor
+	var/capacitor_charge_constant = 10
+	var/list/capacitors
+	var/max_capacitors = 2
+	var/charging = FALSE
+	var/charge_damage_multiplier = 0.5
+
+/obj/item/gun/energy/capacitor/Destroy()
+	if(capacitors)
+		QDEL_NULL_LIST(capacitors)
+	. = ..()
+
+/obj/item/gun/energy/capacitor/Initialize()
+	var/capacitor_volume = 0
+	for(var/i = 1 to max_capacitors)
+		var/obj/item/stock_parts/capacitor/capacitor = new capacitor_type(src)
+		capacitor_volume = min(capacitor_volume, capacitor.max_charge)
+		LAZYADD(capacitors, capacitor)
+	. = ..()
+
+/obj/item/gun/energy/capacitor/afterattack(atom/A, mob/living/user, adjacent, params)
+	. = !charging && ..()
+	
+/obj/item/gun/energy/capacitor/attackby(obj/item/W, mob/user)
+	if(!charging)
+
+		if(isScrewdriver(W))
+			if(length(capacitors))
+				var/obj/item/capacitor = capacitors[1]
+				capacitor.dropInto(user.loc)
+				user.put_in_hands(capacitor)
+				LAZYREMOVE(capacitors, capacitor)
+			else if(power_supply)
+				power_supply.dropInto(loc)
+				user.put_in_hands(power_supply)
+				power_supply = null
+			else
+				to_chat(user, SPAN_WARNING("\The [src] does not have a cell or capacitor installed."))
+				return TRUE
+			playsound(loc, 'sound/items/Screwdriver2.ogg', 25)
+			update_icon()
+			return TRUE
+
+		if(istype(W, /obj/item/cell))
+			if(power_supply)
+				to_chat(user, SPAN_WARNING("\The [src] already has a cell installed."))
+			else if(user.unEquip(W, src))
+				power_supply = W
+				to_chat(user, SPAN_NOTICE("You fit \the [W] into \the [src]."))
+			return TRUE
+
+		if(istype(W, /obj/item/stock_parts/capacitor))
+			if(length(capacitors) >= max_capacitors)
+				to_chat(user, SPAN_WARNING("\The [src] cannot fit any additional capacitors."))
+			else if(user.unEquip(W, src))
+				LAZYADD(capacitors, W)
+				to_chat(user, SPAN_NOTICE("You fit \the [W] into \the [src]."))
+			return TRUE
+
+	. = ..()
+
+/obj/item/gun/energy/capacitor/attack_self(var/mob/user)
+	if(charging)
+		for(var/obj/item/stock_parts/capacitor/capacitor in capacitors)
+			capacitor.charge = 0
+		update_icon()
+		charging = FALSE
+		return TRUE
+	. = ..()
+
+/obj/item/gun/energy/capacitor/proc/charge(var/mob/user)
+	. = FALSE
+	var/tried_charge = FALSE
+	if(!charging && istype(user))
+		charging = TRUE
+		while(!QDELETED(user) && length(capacitors) && charging && user.get_active_hand() == src)
+			var/charged = TRUE
+			for(var/obj/item/stock_parts/capacitor/capacitor in capacitors)
+				if(capacitor.charge < capacitor.max_charge)
+					charged = FALSE
+					tried_charge = TRUE
+					var/use_charge_cost = min(charge_cost * capacitor.rating, round((capacitor.max_charge - capacitor.charge) / capacitor_charge_constant))
+					if(power_supply.use(use_charge_cost))
+						capacitor.charge(use_charge_cost * capacitor_charge_constant)
+						update_icon()
```

Comment:
Do we really need to be doing this 10 times a second? Could we use the obj or the fast processing SS instead?

Reply:
As discussed on Discord, using processing would be a pain. I've lowered the number of times this will fire each second though.